### PR TITLE
[POC] Model Driven SSO Configuration POC 

### PR DESCRIPTION
**Pull request by whitej6**
This is a really old piece of work and is by no means ready for review. Wanted to submit the PR as a jumping off point on using database objects and/or constant config to back SSO configuration.

### CODE REVIEW

File: nautobot/core/settings.py [hunk 1]
```diff
@@ -440,6 +442,8 @@
 CONSTANCE_DATABASE_PREFIX = "constance:nautobot:"
 CONSTANCE_IGNORE_ADMIN_VERSION_CHECK = True  # avoid potential errors in a multi-node deployment
 
+CONSTANCE_MANAGE_SSO = is_truthy(os.getenv("NAUTOBOT_CONSTANCE_MANAGE_SSO", False))
```
Comment (jathanism) [inline nautobot/core/settings.py:445]: Terse docstring here please.

File: nautobot/core/settings_funcs.py [hunk 2]
```diff
@@ -86,3 +90,25 @@ def parse_redis_connection(redis_database):
...
+                    inspect.isclass(obj)
+                    and hasattr(obj, "name")
+                    and obj.name
+                    and backend_module.__spec__.name == obj.__module__
+                ):
+                    available_backends.add(f"social_core.backends.{backend}.{name}")
+        except ModuleNotFoundError:
+            pass
+        except ImportError:
+            pass
```
Comment (jathanism) [inline nautobot/core/settings_funcs.py:113]: Would it be possible for you you use [`django.utils.module_loading.import_string`](https://docs.djangoproject.com/en/4.1/ref/utils/#django.utils.module_loading.import_string) here instead? It does this already but with less code.

We also have [`nautobot.extras.plugin.utils.import_object`](https://github.com/nautobot/nautobot/blob/develop/nautobot/extras/plugins/utils.py#L19), but I hate that it lives inside of `plugin.utils`...

File: nautobot/core/sso.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+from nautobot.users.models import SSOBackend
+
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class NautobotStrategy(DjangoStrategy):
+    """Subclassing to allow for constance config exposed via Nautobot."""
```
Comment (jathanism) [inline nautobot/core/sso.py:19]: Can you please revise this docstring to describe how it is used? It's not clear act a `Strategy` is.

File: nautobot/core/sso.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+    def url_to_promie(self, variable_name, value):
+        """Force text on URL named settings that are instance of Promise."""
```
Comment (jathanism) [inline nautobot/core/sso.py:22]: Why would we want to do this?

File: nautobot/users/admin.py [hunk 5]
```diff
@@ -128,6 +135,30 @@ class TokenAdmin(NautobotModelAdmin):
...
+            "name",
+            "backend",
+            "enabled",
+            "configuration",
+        ]
+
+
+@admin.register(SSOBackend)
+class SSOBackendAdmin(NautobotModelAdmin):
+    form = SSOBackendAdminForm
```
Comment (jathanism) [inline nautobot/users/admin.py:157]: Is there really a need for a custom form? The fields can limited in the `ModelAdmin`.
```suggestion


@admin.register(SSOBackend)
class SSOBackendAdmin(NautobotModelAdmin):
    fields = ["name", "backend", "enabled", "configuration"]
```

File: nautobot/users/models.py [hunk 6]
```diff
@@ -256,3 +257,43 @@ def list_constraints(self):
...
             return [self.constraints]
         return self.constraints
+
+
+#
+# SSO Backend
+#
+
+
+BACKEND_CHOICES = [(backend, backend.split(".")[-1]) for backend in settings.AUTHENTICATION_BACKENDS[:-1]]
```
Comment (jathanism) [inline nautobot/users/models.py:267]: Please to make be define in `choices` enum. 🪨

File: nautobot/users/models.py [hunk 7]
```diff
@@ -256,3 +257,43 @@ def list_constraints(self):
...
+
+class SSOBackend(BaseModel):
+    name = models.CharField(max_length=100, unique=True)
+    backend = models.CharField(max_length=100, choices=BACKEND_CHOICES, unique=True)
+    _backend_name = models.CharField(max_length=100, editable=False)
```
Comment (jathanism) [inline nautobot/users/models.py:272]: Why prefix with `_`?

File: nautobot/users/models.py [hunk 8]
```diff
@@ -256,3 +257,43 @@ def list_constraints(self):
...
+        """Returns the variable namespace for a backend."""
+        return "SOCIAL_CORE_" + self._backend_name.upper().replace("-", "_") + "_"
+
+    def save(self, *args, **kwargs):
+        """Overload save to set _backend_name to provide filtering & help performance."""
+        self._backend_name = import_string(self.backend).name
+        return super().save(*args, **kwargs)
+
+
+SSO_CONFIG_TYPE_CHOICES = [("text", "Text"), ("secret", "Secret")]
```
Comment (jathanism) [inline nautobot/users/models.py:292]: Enum pls. 🙈

**Review by jathanism — CHANGES_REQUESTED** [review]

**Comment by bryanculver** [comment]
Going to squash this into a prototype branch, it's a take at #2547 along with #2549.

Also easier facilitation of multiple SSO backends.

**Pull request merged**